test/run-check-status.bats

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env bats
+
+load test_helper
+
+# Positive assertions: each of these should succeed
+@test "basic return-code checking" {
+  run      true
+  run =0   true
+  run '!'  false
+  run =1   false
+  run =3   exit 3
+  run =5   exit 5
+  run =111 exit 111
+  run =255 exit 255
+  run =127 /no/such/command
+}
+
+# Negative assertions.
+# We can't just run 'bats' on an entire list of tests, because each test here
+# is expected to fail and we need to know _how_ it should fail.
+# So this is a test helper. It assumes that the caller has defined 'commands'
+# as an array of one or more instructions; Those will be written to a .bats
+# tmpfile which we then run. Our function arguments are patterns to search for
+# in the output from the test.
+function _run_test() {
+  # Write a test file
+  local testfile=${BATS_TMPDIR}/bats-testfile-$$.bats
+  rm -f $testfile
+  echo "@test \"negative test\" {"   >$testfile
+  for c in "${commands[@]}"; do
+      echo "    $c"                 >>$testfile
+  done
+  echo "}"                          >>$testfile
+
+  # Run it
+  export BATS_PS1='run> '
+  run '=1' bats $testfile
+  rm -f $testfile
+
+  if [[ "${lines[0]}" != "1..1" ]]; then
+    die "Internal error: expected '1..1' as first line, got '${lines[0]}'"
+  fi
+  if [[ "${lines[1]}" != "not ok 1 negative test" ]]; then
+    die "Internal error: expected 'not ok 1 negative test' as second line, got '${lines[1]}'"
+  fi
+
+  local lineno=1
+  for e in "$@"; do
+    while :; do
+      lineno=$(( lineno + 1 ))
+      if [ $lineno -gt ${#lines[*]} ]; then
+        die "$testname: did not find '$e' in output from this test" \
+            "${lines[@]}"
+      fi
+
+      if [[ ${lines[$lineno]} =~ $e ]]; then
+        lines[$lineno]="${lines[$lineno]}   [MATCHED]"
+        break
+      fi
+    done
+  done
+}
+
+@test "run: unexpected fail" {
+  local -a commands=("run =0 false")
+  _run_test "\(in test file .*, line 2\)" \
```

Comment:
I find `_run_test` quite involved. The usual way of doing this would be extra files for each test that can then be called via `bats testfile`. I am not sure if this solution will be harder to maintain. It definitely requires more understanding than simply running a file with bats. @sublimino What do you think about this approach?

Reply:
One of my earlier iterations had a set of `fixtures` tests (for `xpect`). What I was trying to solve here was being able to have the commands and expected output very close together, for maintainability. I could not find a good way to do so using individual fixture test files; but it's very likely I may have missed something.

Reply:
Well, the code in question is in a separate file but the bats output still shows what went wrong, so I feel there should be enough context to see what is going on.